### DISCUSSION TITLE
Adding nodal constraint equation

### DISCUSSION THREAD

**Discussion by jessecarterMOOSE** [idaholab/moose#21500] [Q&A General] [answered]
Does MOOSE have a system in place like the [linear constraint system in Abaqus](https://classes.engineering.wustl.edu/2009/spring/mase5513/abaqus/docs/v6.6/books/usb/default.htm?startat=pt08ch28s02aus104.html)  that would allow specific variables at specific nodes sum to zero? Might not necessarily need to be a MOOSE Constraint constraint, but maybe this could be handled with a kernel?

Specifically I'd like to apply this to mesh displacement variables rather than applying typical displacement boundary conditions.

**Comment by GiudGiud** [comment]
Hello

My guess is no, seems like it would be an object that would contribute to many different variable residuals and it's not commonly done. 
But it s totally possible and would not be too too difficult, just need to rewrite the computeJacobian and computeResidual routines. 

@lindsayad to make sure

Guillaume

**Comment by lindsayad** [comment, marked as answer]
I don't think we currently have an object as generic as you describe, but we definitely have systems that would support it. Either `NodalConstraints` or `NodalKernels` should be able to do the job

**Reply by jessecarterMOOSE** [reply]
Thanks. I think I might be able to make it work with NodalKernels. How does the theory for that system compare with regular Kernels, i.e. weak form and Jacobians and such?

**Reply by lindsayad** [reply]
`NodalKernels` were originally created for equations without spatial derivative terms (they could have d/dt terms) but for which the variable was still a spatially varying field. Because there are no spatial derivative terms (at least in the use case the system was originally intended for), there are no weak forms. The equations you are solving for presumably depend on the nonlinear degrees of freedom, so you still have Jacobians
